### PR TITLE
Fix skipped tests due to sklearn

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Tests are skipped on master: https://github.com/Lightning-AI/lightning/actions/runs/3317863655/jobs/5481178929#step:22:722

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @carmocca @akihironitta @tchaton @rohitgr7

### CODE REVIEW

**Comment by rohitgr7** [comment]
running now: https://github.com/Lightning-AI/lightning/actions/runs/3322693160/jobs/5492076875#step:22:722

File: tests/tests_pytorch/helpers/datamodules.py [hunk 1]
```diff
@@ -20,7 +20,7 @@
 from pytorch_lightning.core.datamodule import LightningDataModule
 from tests_pytorch.helpers.datasets import MNIST, SklearnDataset, TrialMNIST
 
-_SKLEARN_AVAILABLE = RequirementCache("sklearn")
+_SKLEARN_AVAILABLE = RequirementCache("scikit-learn")
```
Comment (awaelchli) [inline tests/tests_pytorch/helpers/datamodules.py:23]: This is reverting #14475 which was reverting already the other way. 
Either we agree to check for the module, or we agree to check for the package. These two things can have different names. RequirementCache is for the package, module_available is for the module. Correct @carmocca ?

Comment (awaelchli) [inline tests/tests_pytorch/helpers/datamodules.py:23]: @carmocca It might be worth checking that this mistake was not made for other changes in #14475. 
My understanding is that RequirementCache is for packages, while we should use module_available for modules.

**Review by carmocca — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

File: tests/tests_pytorch/helpers/datamodules.py (hunk 1 above)
Comment (carmocca) [reply tests/tests_pytorch/helpers/datamodules.py:23]: Thanks for checking!

File: tests/tests_pytorch/helpers/datamodules.py [hunk 2]
```diff
@@ -54,7 +54,7 @@ def test_dataloader(self):
 class SklearnDataModule(LightningDataModule):
     def __init__(self, sklearn_dataset, x_type, y_type, batch_size: int = 10):
         if not _SKLEARN_AVAILABLE:
-            pytest.skip("`sklearn` is not available.")
+            pytest.skip("`scikit-learn` is not available.")
```
Comment (carmocca) [inline tests/tests_pytorch/helpers/datamodules.py:57]: This will include the reason

```suggestion
            pytest.skip(str(_SKLEARN_AVAILABLE)):
```

File: tests/tests_pytorch/helpers/datamodules.py (hunk 1 above)
Comment (carmocca) [inline tests/tests_pytorch/helpers/datamodules.py:23]: GPU tests timed out. We should check what's the impact of running these tests on CI

**Review by Borda — APPROVED** [review]

File: tests/tests_pytorch/helpers/datamodules.py (hunk 1 above)
Comment (rohitgr7) [reply tests/tests_pytorch/helpers/datamodules.py:23]: checking!

File: tests/tests_pytorch/strategies/test_ddp.py [hunk 3]
```diff
@@ -18,42 +18,37 @@
...
-from tests_pytorch.helpers.simple_models import ClassificationModel
 
 
 @RunIf(min_cuda_gpus=2, standalone=True)
 def test_multi_gpu_model_ddp_fit_only(tmpdir):
-    dm = ClassifDataModule()
-    model = ClassificationModel()
-    trainer = Trainer(default_root_dir=tmpdir, max_epochs=1, accelerator="gpu", devices=2, strategy="ddp")
-    trainer.fit(model, datamodule=dm)
+    model = BoringModel()
```
Comment (carmocca) [inline tests/tests_pytorch/strategies/test_ddp.py:30]: Proposal unrelated to the sklearn change or the standalone failures: Let's just delete these tests? They basically test nothing other than that DDP can launch.

cc @awaelchli

Comment (awaelchli) [reply tests/tests_pytorch/strategies/test_ddp.py:30]: They probably can be replaced with a simpler launcher test. But they are here for a reason, from a time when Lightning was borked when calling multiple trainer.x methods in a row. They act as a regression test in that sense.

Comment (rohitgr7) [reply tests/tests_pytorch/strategies/test_ddp.py:30]: yes, I was thinking too to remove them, but they fail on the ci, so was afraid to just remove them directly.

**Comment by rohitgr7** [comment]
raised an issue regarding the failing deepspeed tests: https://github.com/microsoft/DeepSpeed/issues/2449

File: tests/tests_pytorch/core/test_datamodules.py [hunk 4]
```diff
@@ -236,12 +242,12 @@ def test_full_loop(tmpdir):
     # validate
     result = trainer.validate(model, dm)
     assert dm.trainer is not None
-    assert result[0]["val_acc"] > 0.7
```
Comment (awaelchli) [inline tests/tests_pytorch/core/test_datamodules.py:239]: Do you think we should revert these numbers but instead find another seed that satisfies the existing threshold?

File: tests/tests_pytorch/strategies/test_deepspeed_strategy.py [hunk 5]
```diff
@@ -688,6 +688,7 @@ def test_deepspeed_multigpu_stage_3_manual_optimization(tmpdir, deepspeed_config
     _assert_save_model_is_equal(model, tmpdir, trainer)
 
 
+@pytest.mark.skip()
```
Comment (awaelchli) [inline tests/tests_pytorch/strategies/test_deepspeed_strategy.py:691]: Let's add a reason, pointing to the issue that you created (thanks for that!)

File: tests/tests_pytorch/run_standalone_tests.sh [hunk 6]
```diff
@@ -17,7 +17,7 @@ set -e
 
 # Batch size for testing: Determines how many standalone test invocations run in parallel
 # It can be set through the env variable PL_STANDALONE_TESTS_BATCH_SIZE and defaults to 6 if not set
-test_batch_size="${PL_STANDALONE_TESTS_BATCH_SIZE:-6}"
+test_batch_size="${PL_STANDALONE_TESTS_BATCH_SIZE:-1}"
```
Comment (awaelchli) [inline tests/tests_pytorch/run_standalone_tests.sh:20]: reminder to revert

File: .azure/gpu-tests-pytorch.yml [hunk 7]
```diff
@@ -121,13 +121,6 @@ jobs:
       workingDirectory: src
       displayName: 'Testing: PyTorch doctests'
 
-    - bash: python -m coverage run --source pytorch_lightning -m pytest --ignore benchmarks -v --junitxml=$(Build.StagingDirectory)/test-results.xml --durations=50
```
Comment (awaelchli) [inline .azure/gpu-tests-pytorch.yml:124]: reminder to revert

**Review by awaelchli — APPROVED** [review]

File: tests/tests_pytorch/core/test_datamodules.py (hunk 4 above)
Comment (rohitgr7) [reply tests/tests_pytorch/core/test_datamodules.py:239]: I think it's fine since these are not benchmarks...

File: tests/tests_pytorch/strategies/test_deepspeed_strategy.py (hunk 5 above)
Comment (rohitgr7) [inline tests/tests_pytorch/strategies/test_deepspeed_strategy.py:691]:
```suggestion
@pytest.mark.skip(reason="skipped due to deepspeed/#2449, keep track @rohitgr7")
```

File: tests/tests_pytorch/run_standalone_tests.sh (hunk 6 above)
Comment (rohitgr7) [inline tests/tests_pytorch/run_standalone_tests.sh:20]:
```suggestion
test_batch_size="${PL_STANDALONE_TESTS_BATCH_SIZE:-6}"
```

File: .azure/gpu-tests-pytorch.yml [hunk 8]
```diff
@@ -121,13 +121,6 @@ jobs:
       workingDirectory: src
       displayName: 'Testing: PyTorch doctests'
```
Comment (rohitgr7) [inline .azure/gpu-tests-pytorch.yml:123]:
```suggestion

       - bash: python -m coverage run --source pytorch_lightning -m pytest --ignore benchmarks -v --junitxml=$(Build.StagingDirectory)/test-results.xml --durations=50
         env:
           PL_RUN_CUDA_TESTS: "1"
         workingDirectory: tests/tests_pytorch
         displayName: 'Testing: PyTorch standard'
         timeoutInMinutes: "35"
```

File: .azure/gpu-tests-pytorch.yml [hunk 9]
```diff
@@ -127,7 +127,6 @@ jobs:
       workingDirectory: tests/tests_pytorch
       displayName: 'Testing: PyTorch standard'
       timeoutInMinutes: "35"
```
Comment (rohitgr7) [inline .azure/gpu-tests-pytorch.yml:129]:
```suggestion
      timeoutInMinutes: "35"

```

**Review by akihironitta — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

File: tests/tests_pytorch/accelerators/test_hpu.py [hunk 10]
```diff
@@ -76,6 +77,7 @@ def test_all_stages(tmpdir, hpus):
 
 
 @RunIf(hpu=True)
+@mock.patch.dict(os.environ, os.environ.copy())
```
Comment (awaelchli) [inline tests/tests_pytorch/accelerators/test_hpu.py:80]: Had to set this. On this machine, somehow numpy/sklearn is setting these env vars for MKL:
'KMP_DUPLICATE_LIB_OK', 'KMP_INIT_AT_FORK' and our fixture detected the leakage. No big deal, we just make sure the env remains unchanged after the test.

File: tests/tests_pytorch/callbacks/test_quantization.py [hunk 11]
```diff
@@ -109,7 +109,7 @@ def test_quantize_torchscript(tmpdir):
     trainer = Trainer(callbacks=[qcb], default_root_dir=tmpdir, max_epochs=1)
     trainer.fit(qmodel, datamodule=dm)
 
-    batch = iter(dm.test_dataloader()).next()
+    batch = next(iter(dm.test_dataloader()))
```
Comment (awaelchli) [inline tests/tests_pytorch/callbacks/test_quantization.py:112]: Isn't it strange that we had to fix this in this PR? This failed in the PyTorch 1.13 jobs. Surprised that `.next()` even worked, have never seen that.

**Comment by awaelchli** [comment]
@rohitgr7 I think I got all the failures that were remaining. This should be good to go now.

File: tests/tests_pytorch/callbacks/test_quantization.py (hunk 11 above)
Comment (rohitgr7) [reply tests/tests_pytorch/callbacks/test_quantization.py:112]: yep, because these tests were not running before.

**Comment by rohitgr7** [comment]
thanks @awaelchli for fixing the remaining testt.

File: tests/tests_pytorch/callbacks/test_quantization.py (hunk 11 above)
Comment (awaelchli) [reply tests/tests_pytorch/callbacks/test_quantization.py:112]: Do you know how they were skipped? The only marker on that test is quantization=True and I don't see a connection to the changes in this PR

Comment (rohitgr7) [reply tests/tests_pytorch/callbacks/test_quantization.py:112]: yep: https://github.com/Lightning-AI/lightning/blob/773cb3e8c8d08ea373638d40be0831e8b28ed937/tests/tests_pytorch/helpers/datamodules.py#L112-L115

Comment (rohitgr7) [reply tests/tests_pytorch/callbacks/test_quantization.py:112]: should we add sklearn to runif?

Comment (awaelchli) [reply tests/tests_pytorch/callbacks/test_quantization.py:112]: Oh wow, I smell that this is a bad practice to hide a pytest.skip like this. 

> should we add sklearn to runif?

If we have enough usages that justify it, yes I think so. Otherwise, at least we should put the skip mark on the test itself.

Comment (rohitgr7) [reply tests/tests_pytorch/callbacks/test_quantization.py:112]: yep agreed! component should raise an error, pytest should handles skips
